### ISSUE TITLE
[Feature] Plugin API to include code into generated autoload file(s)

### ISSUE THREAD

**Issue by helhum** [composer/composer#5514]
There are several use cases that need PHP code to be generated at `composer install` time, which will then be executed when including the vendor/autoload.php file.

Currently composer plugins have two possibilities to achieve this and both have been seen in the wild.
1. Add files that will then be included via the autoloading mechanism (e.g. generating an include file which resides in the directory of the plugin package).
2. Add code to vendor/autoload.php by parsing and re-writing the code generated by composer

Both ways are hacky and have their own set of disadvantages and are also likely to break in various scenarios.

I propose to add an additional method / interface to the plugin API (I haven't thought of a concrete implementation yet) that let's composer plugins return an array of files (relative to vendor dir), that will be included when vendor/autoload.php is included.

Thus, when composer generates the autoload files, it adds code that iterates over the array and includes files that are provided by such plugins.

In case this idea fits in the plans of the composer project, I'm willing to implement (and maintain) this feature.

What do you think?

**Comment by Ocramius** [comment]
> I propose to add an additional method / interface to the plugin API (I haven't thought of a concrete implementation yet) that let's composer plugins return an array of files (relative to vendor dir), that will be included when vendor/autoload.php is included.

How about letting the plugin return a list of paths that were generated/affected, then let composer decide what to do (re-generate autoloader or such)?

**Comment by alcohol** [comment]
> There are several use cases that need PHP code to be generated at composer install time, which will then be executed when including the vendor/autoload.php file.

Such as?

**Comment by helhum** [comment]
> Such as?

I saw this question coming ;)

Here are some examples:

https://github.com/Ocramius/PackageVersions
https://github.com/puli/composer-plugin/blob/master/src/PuliPluginImpl.php#L144-L145
https://github.com/TYPO3/CmsComposerInstallers/blob/master/src/Plugin/Core/AutoloadConnector.php#L86

Two of my plugins:
https://github.com/helhum/dotenv-connector/blob/master/src/Plugin.php#L74-L75
https://github.com/TYPO3-Console/typo3-console-plugin/blob/master/src/IncludeFileWriter.php#L79-L80


Basically it is about exposing information that is available at composer install/update build time, that is useful/ needed during application runtime.

1. Package versions
1. Installed packages
1. Installation type
1. environment configuration
1. composer root directory

There might be more. These are the only things I stumbled upon (and used/ added myself).

If there are other ways to achieve this, I'm glad to hear! Otherwise, I'd be glad to help to integrate such a feature.

**Comment by Ocramius** [comment]
Even just stuff like Doctrine ORM/ODM/etc proxies could be generated and autoloaded via classmap this way...

**Comment by Seldaek** [comment]
What speaks against just generating the classes in the paths they are needed and having the autoload config already in the plugin package? I am not entirely sure I understand why you need to add new paths. Got a concrete example of this?

Aside from that, from an implementation perspective, random thoughts: I guess we could pass more info and allow collection of info in PRE_AUTOLOAD_DUMP, althoguh that's a script event right now which makes it a bit difficult, would be easier to change that or if that causes issues introduce a new event maybe after all the autoload config is generated but before it's dumped to files so the plugins can modify that or maybe safer to only allow adding as you initially suggested.

**Comment by helhum** [comment]
> What speaks against just generating the classes in the paths they are needed and having the autoload config already in the plugin package

Any package (including a plugin package) can be pulled into a project using a repository of type "path". This means, that the package will be symlinked into the project directory. If the plugin generates a file that either intends to expose paths from the project or access other files from the project, it won't work. So the generated files must be in a place relative to the project dir (not within a symlink)

Same issue btw. currently exists with binaries and packages pulled in via "path". The binaries won't work if they are just symlinked.

**Comment by Seldaek** [comment]
> Any package (including a plugin package) can be pulled into a project using a repository of type "path". This means, that the package will be symlinked into the project directory.

Is that so? Do you have an easy way to reproduce this?

> If the plugin generates a file that either intends to expose paths from the project or access other files from the project, it won't work. So the generated files must be in a place relative to the project dir (not within a symlink)

Not sure what you mean by relative to the project dir, but I assume you mean outside the vendor dir (which is where symlinked path repo would live, much like regular packages). I also don't know what you mean by `expose paths from the project or access other files from the project`. It would really help if you linked to an actual example with code because I am afraid words are failing here :) I'd just like to understand the use case fully before discussing further.

**Comment by sbuzonas** [comment]
> Any package (including a plugin package) can be pulled into a project using a repository of type "path". This means, that the package will be symlinked into the project directory.

Repositories are root only configuration, and primarily just a development workflow problem IMO.  That being said, what is wrong with the path repository options `"symlink": false`?

> Same issue btw. currently exists with binaries and packages pulled in via "path". The binaries won't work if they are just symlinked.

How are you referencing paths that symlinks aren't working for you?  I've never had a problem with bin files symlinked into a symlinked path repository.

**Comment by helhum** [comment]
> Currently composer plugins have two possibilities to achieve this and both have been seen in the wild.

Meanwhile I found a third possibility, which completely covers my use case, which is adding include sections to the autoload definition of the root package during PRE_AUTOLOAD_DUMP https://github.com/helhum/dotenv-connector/blob/master/src/Plugin.php#L87-L90

To answer your questions @sbuzonas:

> Repositories are root only configuration, and primarily just a development workflow problem IMO. 

uhm, no not entirely. Imho they are useful in monorepos as well.

> That being said, what is wrong with the path repository options "symlink": false?

Nothing. It is just that maybe composer should/ could detect that case and just do not symlink. or just never symlink the binaries but create files.

> How are you referencing paths that symlinks aren't working for you? I've never had a problem with bin files symlinked into a symlinked path repository.

The symlinks work of course, but detecting the autoload.php file does not work any more. The script assumes to be located below vendor/vendorName/packageName, but in fact is located in any other path (specified in the repository).


Anyway. I found a very good solution for my initial use case, thus will close the feature request.

Thanks to everybody taking the time for participating in the discussion.
